### PR TITLE
Max record bug fix

### PR DESCRIPTION
**Pull request by juggernautsei**
This is to fix the inserting of patients in a table that has more than 10000 records.

### CODE REVIEW

File: interface/new/new_comprehensive_save.php [hunk 1]
```diff
@@ -28,11 +28,9 @@
 // other interfaces can still read the data during this lock, however
 // sqlStatement("lock tables patient_data read");
 
-$result = sqlQuery("SELECT MAX(pid)+1 AS pid FROM patient_data");
```
Comment (robertdown) [inline interface/new/new_comprehensive_save.php]: Do we know why this was failing? I just tried that SQL and it returned the expected results

Comment (juggernautsei) [inline interface/new/new_comprehensive_save.php]: How many records are in the table you executed this query on?

Comment (juggernautsei) [inline interface/new/new_comprehensive_save.php]: This query does not fail, what happens it that it maxes out at 10000.

Comment (robertdown) [inline interface/new/new_comprehensive_save.php]: I have far fewer than 10000 records, however I manually updated my last row's PID to 10000. When I ran the query it gave back 10001. I can try with 10000 actual records, standby

File: interface/new/new_comprehensive_save.php [hunk 2]
```diff
@@ -28,11 +28,9 @@
...
-
+$result = sqlQuery("SELECT id AS pid FROM patient_data ORDER BY id DESC LIMIT 1");
```
Comment (bradymiller) [inline interface/new/new_comprehensive_save.php]: should not be using id here. it should be:
``````
$result = sqlQuery("SELECT pid FROM patient_data ORDER BY id DESC LIMIT 1");
``````
(Note you also have 2 unrelated files with changes in this commit; not a biggie, which we can deal with when/if bring in this fix)

Comment (juggernautsei) [inline interface/new/new_comprehensive_save.php]: Help me out with the unrelated files. I don't see them when I do my commit. Even in the first commit, the dialog only showed me one file.

POST git-receive-pack (612 bytes)
Pushing to https://github.com/juggernautsei/openemr.git
To https://github.com/juggernautsei/openemr.git
   180cb6e..78ef24d  max-record-bug -> max-record-bug
updating local tracking ref 'refs/remotes/origin/max-record-bug'

Just trying to figure out how more files are slipping by without me seeing them in the dialog box.

**Comment by bradymiller** [comment]
@robertdown ,
This has been lingering for awhile and is still an issue on Sherwin's side. My guess is this may be a windows issue. The solution does not really compromise anything, so seems ok to bring in. Just want to get your ok before I bring this in (plan to bring in manually to only bring in the change to the interface/new/new_comprehensive_save.php file).
thanks,
-brady

**Comment by juggernautsei** [comment]
When I looked up MAX(), the way that it is being used in the code is incorrect according to http://php.net/manual/en/function.max.php
The max function does not get the last record in the table. It is used to compare an array.

**Comment by robertdown** [comment]
@juggernautsei the MAX function is a SQL function here, not a PHP function. See https://dev.mysql.com/doc/refman/5.5/en/group-by-functions.html#function_max

@bradymiller let me take a look one last time and then we can do this

**Comment by mdsupport** [comment]
Never heard of a a bug with MAX of mysql.  The only problem with its use I see here is when inserting first record, MAX would be NULL.  NULL+1 is NULL.  So you would need either : 

1. Change sql as `SELECT (IFNULL(MAX(pid),0)+1) AS pid FROM patient_data`  (prefer) or
2. In php code check for null result and assign 1.

**Comment by juggernautsei** [comment]
Guys your are looking in the wrong direction. The MAX(id) does bring back a result 10000 seems to be as high as it will go. So, what fails is that 10000 is already in use so it can't insert the record. So the insert fails. Notice how the PID is set to 10000 so the MAX(id) returns a value but it is incorrect. It should have been 21644.
Review the orginal error
"ERROR: insert failed: INSERT INTO patient_data SET pid = '**10000**', date = NOW(),
title = 'Mr.', fname = 'Patient', mname = '', lname = 'Demo', pubpid =
'**10000**', DOB = '2016-08-25', sex = 'Male', ss = '', drivers_license =
'', status = '', genericname1 = '', genericval1 = '', genericname2 = '',
genericval2 = '', billing_note = '', street = '', city = '', state =
'', postal_code = '', country_code = '', county = '', mothersname = '',
guardiansname = '', contact_relationship = '', phone_contact = '',
phone_home = '', phone_biz = '', phone_cell = '', email = '',
email_direct = '', providerID = '', ref_providerID = '', pharmacy_id =
'0', hipaa_notice = '', hipaa_voice = '', hipaa_message = '', hipaa_mail
= '', hipaa_allowsms = '', hipaa_allowemail = '', allow_imm_reg_use = '',
allow_imm_info_share = '', allow_health_info_ex = '', allow_patient_portal
= '', care_team = '', cmsportal_login = '', occupation = '', industry =
'', language = '', ethnicity = '', race = '', financial_review = '',
family_size = '', monthly_income = '', homeless = '', interpretter = '',
migrantseasonal = '', referral_source = '', vfc = '', religion = '',
deceased_date = '', deceased_reason = ''

**Comment by mdsupport** [comment]
You are correct.  The sql fails but not because of pid.   Here is the full error when the SQL is pasted in mysql query:

> MySQL said: Documentation
> #1366 - Incorrect integer value: '' for column 'providerID' at row 1

providerID is int(11) with default NULL.

**Comment by juggernautsei** [comment]
Ok, this is going nowhere fast. 
Let us go around this bush again. The snippet I gave is not the action item. 

I clearly stated that this only happens when the patient count in the patient_data table is above 10000.

I stated, when I received the above error message. The count should have been 22000+ but the code returned 10000 as the MAX(id). 

Is this clear now ;-)

**Comment by sunsetsystems** [comment]
I question the statement "The MAX(id) does bring back a result 10000 seems to be as high as it will go." This is not normal behavior and I have seen many sites with more than 10,000 clients. Suggest you find out why this is happening in your case. Corrupted index perhaps? Column definition not as expected?

**Comment by robertdown** [comment]
This comment won't immediately solve the issue at hand, but this is a good case for UUID PUDs 😊

**Comment by robertdown** [comment]
Also, @sunsetsystems, any reason the SQL wasn't written as `SELECT pid FROM patient_data ORDER BY pid ASC LIMIT 1`? That would give the same result, no?

**Comment by sunsetsystems** [comment]
No special reason (assuming you mean DESC). Though in this case it's important to find the cause of the problem and not hide it as the solution.

**Comment by juggernautsei** [comment]
I understand. It maybe just those things. Makes sense to me.

**Comment by mdsupport** [comment]
So you are trying to add 22,000+ records to patient_data - (presumably as a batch) and you get error at 10,000 - always exactly 10,000, not 9,998 or 10,005 or anything else?  Are we to assume, if you manually change one of the existing pid to 100,000 (not 10,000), you get error?  Then you try to insert a new record manually, you do not get automatic pid = 100,001?

**Comment by juggernautsei** [comment]
No, not trying to do this in a batch.

Let's take this from the beginning. I imported patient data into the table last year as a startup of using the program. I noticed that when using the demographics page to enter a new patient after the import. It failed at 10,000 because the MAX(id) would only return the number 10000. So I changed the code to get the array of id's limit 1 so I would get the real last digit in the table.
I forgot about it until I did the upgrade from 4.2.x to 5.0.0. The problem returned as before a year earlier.
So, I changed the code back to get the last ID as previously done.

**Comment by sunsetsystems** [comment]
Back up the site. Go into phpmyadmin and look at the structure for patient_data. Make sure it's bigint(20). Then run the query "select max(pid) from patient_data" and see what you get. If it's not right, delete the "pid" index and run the query again (it will be slower). Then you can create the index again.

If still no luck, go to the Operations tab and try Analyze Table and Check Table.

**Comment by juggernautsei** [comment]
I ran the query from the database before doing anything to see if I would get 10000 and I did not. I got 25815. So, we can close this PR?

**Comment by sunsetsystems** [comment]
OK closed.
